### PR TITLE
more vram savings: fused backpass

### PR DESCRIPTION
**Pull request by dxqb**
Fused backpass is another technique to save vram. It is mostly implemented in the host trainer, not the optimizer - but it does need a little support from the optimizer.

Effects are significant:
| Model | Prodigy | + sliced&factored | + fused backpass | AdamW |
|--------|--------|--------|--------|--------|
| Flux Dev NF4 LoRA Rank 128 | 12,1 GB | 9,0 GB | **8,4 GB** | 10.5 GB |
| SDXL bf16 full unet finetune | 31,9 GB | 12,6 GB | **7,9 GB**  | 21,7 GB |
| Flux Dev bf16 full finetune | approx. 144 GB | approx. 53 GB | (**15,1 GB**) | approx. 96 GB |

Fused back pass is a known technique described here:
https://pytorch.org/tutorials/intermediate/optimizer_step_in_backward_tutorial.html

By applying the gradients during the backpass, you can get rid of almost the entire gradient buffer, thereby saving vram.

To support this method, the optimizer has to provide a method called `step_parameter()`, that makes an optimizer step for a single parameter tensor. This is what this PR is about:
Re-factoring the code, so that the calculation of `d` once per optimizer step, and the inner loops that apply the gradients, are separated.

Would you support making these changes?
Making these changes downstream in the training tool would require copying the entire code of Prodigy. I'd prefer to keep the code synced.
This is only a draft, some things could be made nicer. My focus was on not breaking things and changing as little as possible.

The last line in the table above is fine tuning a 12B model in bf16 precision using fused backpass, but also a sophisticated async offloading technique recently developed by @Nerogar. With minimal performance impact, it offloads parts of the model that are currently not used to RAM. Even very large models can be trained on consumer hardware.
It doesn't require anything additional of the optimizer, but it does require the fused backpass described above.
Here is a technical description:
https://github.com/Nerogar/OneTrainer/blob/3963bb2a4eea9a894fa39c151987fa5816da15e2/docs/RamOffloading.md

The other lines in the table are only using fused backpass.

### CODE REVIEW

**Comment by adefazio** [comment]
Would it be possible to keep the original prodigy as ProdigyReference in prodigy_reference.py and write a unit test that compares the behavior to confirm this new version stays the same for standard uses cases? Thanks!
This would also be useful for the adafactor case, to avoid too much code complexity, we want to keep the original prodigy reference somewhere with as simple an implementation as possible so people can modify it and build upon it for research purposes. The main Prodigy version can have all the bells and whistles.

**Comment by dxqbYD** [comment]
reference: sure

what do you have in mind as a unit test? Is there some recognized optimization unit test?
One of the experiments from your paper? Or something simpler? Maybe pytorch has a simple optimization example in their repo.

**Comment by adefazio** [comment]
Just a short function that compares the outputs using random gradients after a few steps, I have something like this for schedule-free to compare the reference implementations against the optimized ones:
https://github.com/facebookresearch/schedule_free/blob/main/schedulefree/test_schedulefree.py

**Comment by Arcitec** [comment]
Wow. It feels like Christmas to see all these optimizations for Prodigy. The greatest optimizer getting great optimizations. ;)

To clarify something regarding having a `prodigy_reference.py` which matches the paper (a good idea for reference), I guess it could contain the 0-tensor VRAM optimization, but not the new slicing parameter which was recently merged? Then that file can stay archived as the original paper reference. :)

**Comment by konstmish** [comment]
I checked the code and compared it to the theoretical method in the paper, both paper and the current implementation use the previous value of `d` to make the update, and we can compute the new value of `d` *after* doing all updates. So I'm positive we can merge something like this, although I'll probably use this PR as a reference rather than merging directly.

So for now the plan is:
1. Add the old implementation just for reference to the repo (#29 but with the fixed fsdp bug).
2. Update the way we do scaling of `exp_avg` and `exp_avg_sq` (remove scaling by `d` following [the discussion](https://github.com/konstmish/prodigy/pull/25#issuecomment-2554791319) of #25) and fix the bias correction (a long-standing issue).
3. Change the implementation to allow for fusing backpass following this PR.
4. Add the AdaFactor version (#25) after testing it on CIFAR10. Maybe add a short pdf to the repo to explain the estimation of `d`.
5. Add the ScheduleFree version. Maybe add a short pdf with a convergence proof to the repo.
6. Look into adding support for multiple parameter groups, I don't have an opinion on this at the moment but it seems like an important option for people fine-tuning diffusion models.

**Comment by dxqbYD** [comment]
> 
> So for now the plan is:
> 

:thumbsup: let us know if we can help with testing or anything else

**Comment by JaheimLee** [comment]
> I checked the code and compared it to the theoretical method in the paper, both paper and the current implementation use the previous value of `d` to make the update, and we can compute the new value of `d` _after_ doing all updates. So I'm positive we can merge something like this, although I'll probably use this PR as a reference rather than merging directly.
> 
> So for now the plan is:
> 
> 1. Add the old implementation just for reference to the repo ([Archive the original paper implementation #29](https://github.com/konstmish/prodigy/pull/29) but with the fixed fsdp bug).
> 2. Update the way we do scaling of `exp_avg` and `exp_avg_sq` (remove scaling by `d` following [the discussion](https://github.com/konstmish/prodigy/pull/25#issuecomment-2554791319) of [more VRAM savings: no first moment and factored second moment #25](https://github.com/konstmish/prodigy/pull/25)) and fix the bias correction (a long-standing issue).
> 3. Change the implementation to allow for fusing backpass following this PR.
> 4. Add the AdaFactor version ([more VRAM savings: no first moment and factored second moment #25](https://github.com/konstmish/prodigy/pull/25)) after testing it on CIFAR10. Maybe add a short pdf to the repo to explain the estimation of `d`.
> 5. Add the ScheduleFree version. Maybe add a short pdf with a convergence proof to the repo.
> 6. Look into adding support for multiple parameter groups, I don't have an opinion on this at the moment but it seems like an important option for people fine-tuning diffusion models.

Hi, maybe you can try [APOLLO](https://github.com/zhuhanqing/APOLLO) and replace AdaFactor with it? @konstmish

**Comment by dxqb** [comment]
Since there is now an RAdam implementation, and requests to have Prodigy work with 2 other optimizers (the one above and ADAPT), I'm wondering if there is a way to implement Prodigy as a learning rate scheduler instead of an optimizer, and therefore have it work with any optimizer?
Just thinking out loud

**Comment by konstmish** [comment]
> Hi, maybe you can try [APOLLO](https://github.com/zhuhanqing/APOLLO) and replace AdaFactor with it?

Thanks for the suggestion, at the moment I see more reasons to add AdaFactor as it's been around for quite a while and still seems to be popular.

> I'm wondering if there is a way to implement Prodigy as a learning rate scheduler instead of an optimizer

That's a good point and I wish it was possible, but one of the main things that Prodigy does is change how the second-order moments are estimated, so it's not just about estimating the learning rate. Even implementing it as a wrapper for other optimizers seems impossible right now since we can't just let the optimzier do its update and then slightly change it, we actually have to use a different update rule.

**Comment by dxqb** [comment]
did you change the update rule for Prodigy compared to Adam? I thought only for schedule-free.
the Prodigy updates rule seems identical, except `dlr` stuff which could be the output of a LR scheduler.

Anyway, still just thinking out loud, I haven't fully thought it through. Feel free to ignore this comment.

**Comment by dxqb** [comment]
implementation of Prodigy with Fused Backpass support:  https://github.com/Koratahiu/Advanced_Optimizers/tree/main
